### PR TITLE
fix: terraform nested docs backToLinkProps

### PR DESCRIPTION
**Pull request by zchsh**
## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Updates "nested" docs categories to have a "back to" link in the sidebar that leads to the parent category, rather than the product home page.

## 🤷 Why

To match designs.

## 🛠️ How

- Splits out a new utility `get-back-to-link` under `views/docs-view/utils` to handle some revised logic
- In `get-back-to-link`, if the current `rootDocsPath` is "nested", looks for the parent `rootDocsPath`, and if it can find one, uses it to build the `backToLinkProps`

### Context

terraform-plugins-sdkv2  update backToLinkProps for plugins-sdkv2

### Preview from this PR

> **Note**: the scope of this PR is to address the "back to" link only. There are other changes we need to make to match designs, these are list as subtasks of 🎟️  [Implement design updates for Terraform](https://app.asana.com/0/1202097197789424/1202663176963860/f) and will be handled separately.

CleanShot 2022-08-16 at 15 33 15@2x

## 🧪 Testing

Confirm that this PR fixes the issue:

- [ ] Visit a "nested" docs category and confirm that the "back to" link leads to the "parent" category
    - For example, on [/terraform/plugin/framework][/terraform/plugin/framework], the "back to" link should lead to [/terraform/plugin][/terraform/plugin]

And ensure that existing back-to links have not been affected:

- [ ] Visit a non-"nested" docs category, and confirm that the "back to" link works as previously
    - For example, on [/vault/docs][/vault/docs], the "back to" link should lead to [/vault][/vault]

[task]: https://app.asana.com/0/1202097197789424/1202785022290151/f
[preview]: https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app
[/terraform/plugin]: https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app/terraform/plugin
[/terraform/plugin/framework]: https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app/terraform/plugin/framework
[/vault]: https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app/vault
[/vault/docs]: https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app/vault/docs

### CODE REVIEW

File: src/views/docs-view/server.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ import {
 
 // Local imports
 import { getProductUrlAdjuster } from './utils/product-url-adjusters'
+import { getBackToLink } from './utils/get-back-to-link'
```
Comment (zchsh) [inline src/views/docs-view/server.ts:26]: I've split out `getBackToLink` to a separate utility file, it felt a bit cleaner that way.

File: src/views/docs-view/utils/get-back-to-link.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * For other cases, the `backToLinkProps` will lead to the product home page.
+ *
+ * A rootDocsPath is considered "nested" if its path contains a slash.
+ * If we have a "nested" rootDocsPath, we should try to find the parent
+ * rootDocsPath. If we can't find a "parent" rootDocsPath,
+ * or if this is not a "nested" rootDocsPath, then we should use
+ * a back-to "<product> Home" link.
+ */
+export function getBackToLink(
```
Comment (zchsh) [inline src/views/docs-view/utils/get-back-to-link.ts:19]: Have leaned hard into the function description here to add context on why this feels necessary. Happy to expand further if anything is unclear 🙇‍♂️

File: src/data/terraform.json [hunk 3]
```diff
@@ -153,7 +153,7 @@
 		},
 		{
 			"iconName": "docs",
-			"name": "Plugin",
+			"name": "Plugin Development",
```
Comment (zchsh) [inline src/data/terraform.json:156]: Have updated this one property to make it clear how we're intending to match design. I have a task to update remaining `rootDocsPaths` icons and names, intent is to handle that separately: 🎟️  [[terraform] update docs sub paths icons and text](https://app.asana.com/0/1202097197789424/1202691941299957/f)

**Comment by zchsh** [comment]
@BRKalow @ashleemboyer This PR sets up some logic for "back to" links in `DocsView` sidebars of nested docs categories (such as [/terraform/plugin/framework](https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app/terraform/plugin/framework), which is a category nested within [/terraform/plugin](https://dev-portal-git-zsfix-terraform-nested-back-to-78210c-hashicorp.vercel.app/terraform/plugin)).

Let me know what ya'll think 🙇‍♂️

**Review by BRKalow — APPROVED** [review]
This makes sense to me. One observation, it feels like we'll also want to update our breadcrumbs as we introduce "nested" sections like this. E.g. when I see "< Plugin Development," I would expect the breadcrumbs to reflect the nesting:

```
Developer > Terraform > Plugin Development > Plugin Framework
```

(meta-observation: we have a [concept of nav "levels"](https://app.asana.com/0/1202097197789424/1202518893776071/f) spread across a few different places now: sidebar nav data, breadcrumbs, back to link 💭)

**Comment by zchsh** [comment]
@BRKalow Ah great call re: breadcrumb bar - I've set up a separate task 🎟️ [[terraform] update breadcrumb links for "nested" docs categories](https://app.asana.com/0/0/1202809779151195/f) to handle that.

Definitely agreed that trying to tackle this from a broader "share nav levels across the app" perspective would be ideal. For now I'll likely keep focusing on getting Terraform polished up, while doing my best to abstract fixes & features to apply to all products where possible. Very much hoping to have bandwidth to do some refactors once the polish is done. Let me know if that sounds like a good plan to ya 🙇‍♂️

File: src/views/docs-view/utils/__tests__/get-back-to-link.test.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
+import { ProductName, ProductSlug } from 'types/products'
+import { getBackToLink } from '../get-back-to-link'
+
+describe('fullPathFromRelativeHref', () => {
```
Comment (ashleemboyer) [inline src/views/docs-view/utils/__tests__/get-back-to-link.test.ts:4]: Little typo here!

**Comment by ashleemboyer** [comment]
This doesn't handle the extra nav level on mobile, is there a new ticket for that?

**Comment by zchsh** [comment]
@ashleemboyer Ah great call, that definitely slipped past me in design review. I've made 🎟️ [[terraform] consider mobile nav design for "nested" docs categories](https://app.asana.com/0/0/1202809171194512/f).

TLDR; I think it'll be quite complex to get merge `nav-data` from multiple docs categories into a single set of mobile nav menu panes. We also don't seem to have specific designs for this (at least as far as I've been able to track down). So in the ticket I've proposed deferring mobile menu refinements for "nested" docs categories to post-GA. Have added some more detailed thoughts and rationale to the ticket.

Let me know what you think! 🙇‍♂️

File: src/views/docs-view/utils/__tests__/get-back-to-link.test.ts (hunk 4 above)
Comment (zchsh) [reply src/views/docs-view/utils/__tests__/get-back-to-link.test.ts:4]: Ah thanks, renamed the function and forgot to rename it here 😅 Will handle in my next PR! 👍
